### PR TITLE
feat: Handle complex/nested data types from a parquet file in Parquet Data Source: Struct, Repeated Struct, Maps and Timestamp of type SimpleGroup

### PR DESCRIPTION
**Pull request by Meghajit**
Solves for #137

### CODE REVIEW

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java [hunk 1]
```diff
@@ -49,14 +49,27 @@ public Object parseSimpleGroup(SimpleGroup simpleGroup) {
...
         if (field != null) {
             inputValues = (List<Long>) field;
         }
         return inputValues.toArray(new Long[]{});
     }
 
+    @Override
+    public Object parseRepeatedSimpleGroupField(SimpleGroup simpleGroup) {
+        String fieldName = fieldDescriptor.getName();
+        ArrayList<Long> deserializedValues = new ArrayList<>();
```
Comment (gauravsinghania) [inline dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java:63]: let's keep variable naming consistent with other classes? longArray?

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/repeated/RepeatedEnumHandler.java [hunk 2]
```diff
@@ -54,7 +55,19 @@ public Object transformFromProto(Object field) {
 
     @Override
     public Object transformFromParquet(SimpleGroup simpleGroup) {
-        return null;
+        String defaultEnumValue = fieldDescriptor.getEnumType().findValueByNumber(0).getName();
+        List<String> deserializedEnumArray = new ArrayList<>();
```
Comment (gauravsinghania) [inline dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/repeated/RepeatedEnumHandler.java:59]: enumArray? For consistency with other similar classes?

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/deserialization/SimpleGroupDeserializer.java [hunk 3]
```diff
@@ -49,19 +48,14 @@ public Row deserialize(SimpleGroup simpleGroup) {
 
     private Row addTimestampFieldToRow(Row row, SimpleGroup simpleGroup, Descriptors.Descriptor descriptor) {
         Descriptors.FieldDescriptor fieldDescriptor = descriptor.findFieldByNumber(timestampFieldIndex);
-        String timestampFieldName = fieldDescriptor.getName();
-        if (SimpleGroupValidation.checkFieldExistsAndIsInitialized(simpleGroup, timestampFieldName)) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/deserialization/SimpleGroupDeserializer.java:53]: Previously, there was only one way in which timestamps could be parsed from a simple group. Also, the parsing logic for transforming a simple group timestamp to row was inlined here.

Now with this story,  there are two ways in which timestamps can be parsed: as int64 AND as a nested simple group.
Logic for which one to choose and how to parse is already present in TimeStampHandler along with these necessary validations. Hence, we just call `TimestampHandler::transformFromParquet()` to get the timestamp row.

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java [hunk 4]
```diff
@@ -1,9 +1,94 @@
...
+     *    }
+     * </pre>
+     * }
+     * The outer group is always repeated. key and value are constant field names.
+     *
+     * @param simpleGroup The SimpleGroup object inside which the map field is present
+     * @param fieldName   The name of the map field
+     * @return true, if the map structure follows the spec and false otherwise.
+     */
+    public static boolean checkIsLegacySimpleGroupMap(SimpleGroup simpleGroup, String fieldName) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java:34]: There are mainly 2 reasons:
1. SRP: Even though it uses map, this function looks and behaves like a validator. It returns a boolean, yes or no answer.  I am giving this class a responsibility of validating simple groups against a field, that's it. Whether it's a map field, enum field, string field or what kind of validation doesn't quite matter. Maybe in the future if we start supporting simple group structs, we can also put validators here on whether it is a struct that is valid.
2. Modularity and Ease of Understanding: If you check this function carefully, it does a quite many `if` based checks and hence this needs to be tested for different corner cases. The unit tests are verbose. Here it is a public function and can be tested as it is, but when placed in MapHandler, this would ideally become a private function and hence the entire unit would need to be tested through the public interface of `transformFromParquet`. This method already has a lot of big tests right now and moving this validator function to MapHandler would result in a lot of hard to understand tests. The MapHandler class is already quite big and I don't want to make it bigger.

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java (hunk 1 above)
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java:63]: Will do this

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/repeated/RepeatedEnumHandler.java (hunk 2 above)
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/repeated/RepeatedEnumHandler.java:59]: will do this

Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/repeated/RepeatedEnumHandler.java]: Fixed via commit https://github.com/odpf/dagger/pull/148/commits/a28474cef99f9a873be2ff405380bd27f2a9f0b7

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java (hunk 1 above)
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/typehandler/primitive/LongHandler.java]: Fixed via commit https://github.com/odpf/dagger/pull/148/commits/a28474cef99f9a873be2ff405380bd27f2a9f0b7

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java:34]: As discussed, let's keep this for now

File: dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java [hunk 5]
```diff
@@ -1,9 +1,94 @@
...
+     *          }
+     * </pre>
+     * }
+     * The validation checks below follow the <a href="https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#maps">Apache Parquet LogicalTypes Specification</a>for Maps.
+     *
+     * @param simpleGroup The SimpleGroup object inside which the map field is present
+     * @param fieldName   The name of the map field
+     * @return true, if the map structure follows the spec and false otherwise.
+     */
+    public static boolean checkIsStandardSimpleGroupMap(SimpleGroup simpleGroup, String fieldName) {
```
[the remark this answers is not in the document]
Comment (Meghajit) [reply dagger-common/src/main/java/io/odpf/dagger/common/serde/parquet/SimpleGroupValidation.java:64]: As discussed, we can keep this for now
